commit 5bb8ba02e38fba786fd32f10c604650783e07ca2 (refs/changes/12/312/1)
Author: Pedro P. Candel <pedro@joyent.com>
Date:   2016-08-18T19:01:18+02:00 (3 years, 2 months ago)
    
    TOOLS-1521 sdcadm should avoid updating duplicated agent instances on the same CN

diff --git a/lib/procedures/update-agent-v1.js b/lib/procedures/update-agent-v1.js
index 3190ebd..c2298fa 100644
--- a/lib/procedures/update-agent-v1.js
+++ b/lib/procedures/update-agent-v1.js
@@ -156,6 +156,34 @@ UpdateAgentV1.prototype.execute = function uaExecute(opts, callback) {
                 });
             },
 
+            /*
+             * TOOLS-1521: This is a corner case which will happen only after
+             * manual copying of an agent into the agents dir, for example when
+             * a backup of an agent directory is created on the same dir.
+             *
+             * This will cause an attempt to update twice the same agent, with
+             * the consequent error reported by sdcadm.
+             *
+             * Given that it's not easy to notify the root cause of this error,
+             * we'll provide this function to give operators a clue of what the
+             * problem is.
+             */
+            function preventDuplicateInstances(_, next) {
+                var instUUIDs = change.insts.map(function (ins) {
+                    return ins.instance;
+                });
+
+                var duplicates = instUUIDs.some(function (ins, pos) {
+                    return (instUUIDs.indexOf(ins) !== pos);
+                });
+
+                if (duplicates.length) {
+                    return next(new UpdateError(format(
+                        'There are duplicated instances for agent %s',
+                        change.insts[0].service)));
+                }
+                next();
+            },
             function checkCNsAvailability(_, next) {
                 sdcadm.cnapi.listServers({
                     uuids: change.insts.map(function (i) {
